pulp_docker/app/tasks/sync_stages.py

```diff
@@ -42,26 +33,103 @@ def __init__(self, remote):
         self.remote = remote
 
     async def run(self):
-        """
-        Build and emit `DeclarativeContent` for each Tag.
-        """
-        log.debug("Fetching tags list for upstream repository: {repo}".format(
-            repo=self.remote.upstream_name
-        ))
-        relative_url = '/v2/{name}/tags/list'.format(name=self.remote.namespaced_upstream_name)
-        tag_list_url = urljoin(self.remote.url, relative_url)
-        list_downloader = self.remote.get_downloader(tag_list_url)
-        await list_downloader.run()
-
-        with open(list_downloader.path) as tags_raw:
-            tags_dict = json.loads(tags_raw.read())
-            tag_list = tags_dict['tags']
-
-        for tag_name in tag_list:
-            tag_dc = self.create_pending_tag(tag_name)
-            await self.put(tag_dc)
-
-    def create_pending_tag(self, tag_name):
+        future_manifests = []
+        list_dc_counter = 0
+        man_dc_counter = 0
+        put_later_blob_dc = []
+        tag_list = []
+        to_download = []
+        man_dcs = {}
+
+        with ProgressBar(message='Downloading tag list for the repo', total=1) as pb:
+            relative_url = '/v2/{name}/tags/list'.format(name=self.remote.namespaced_upstream_name)
+            tag_list_url = urljoin(self.remote.url, relative_url)
+            list_downloader = self.remote.get_downloader(tag_list_url)
+            await list_downloader.run()
+
+            with open(list_downloader.path) as tags_raw:
+                tags_dict = json.loads(tags_raw.read())
+                tag_list = tags_dict['tags']
+                tag_list = ['musl', 'latest', 'glibc', '1-musl', '1-ubuntu', '1.29']
+                #tag_list =['a']
+
+            pb.increment()
+
+        with ProgressBar(message='Creating Download requests for Tags') as pb:
+            # could use total of len(tag_list) but if schema1 is present done will be < total
+            for tag_name in tag_list:
+                relative_url = '/v2/{name}/manifests/{tag}'.format(
+                name=self.remote.namespaced_upstream_name,
+                tag=tag_name,
+                )
+                url = urljoin(self.remote.url, relative_url)
+                downloader = self.remote.get_downloader(url=url)
+                to_download.append(downloader.run(extra_data={'headers': V2_ACCEPT_HEADERS}))
+                pb.increment()
+
+        with ProgressBar(message='Parsing SchemaV2 Tags') as pb:
+            while to_download:
```

Comment:
Why did you go with this pattern rather than asyncio.as_completed(to_download)? This seems fine, just curious.

Reply:
i don't think there was any particular reason, i can change it to as_completed for the sake of readability